kvdb/postgres/db.go

```diff
@@ -0,0 +1,238 @@
+
+	CREATE TABLE IF NOT EXISTS public.`+sequenceTableName+`
+	(
+		table_name TEXT NOT NULL PRIMARY KEY,
+		sequence BIGINT
+	);
+	`+hexCreateTableSql)
+	if err != nil {
+		_ = dbConn.Close()
+
+		return nil, err
+	}
+
+	backend := &db{
+		cfg:               config,
+		prefix:            prefix,
+		ctx:               ctx,
+		db:                dbConn,
+		hexTableName:      hexTableName,
+		sequenceTableName: sequenceTableName,
+	}
+
+	return backend, nil
+}
+
+// getTimeoutCtx gets a timeout context for database requests.
+func (db *db) getTimeoutCtx() (context.Context, func()) {
+	if db.cfg.Timeout == time.Duration(0) {
+		return db.ctx, func() {}
+	}
+
+	return context.WithTimeout(db.ctx, db.cfg.Timeout)
+}
+
+// getPrefixedTableName returns a table name for this prefix (namespace).
+func (db *db) getPrefixedTableName(table string) string {
+	return fmt.Sprintf("%s_%s", db.prefix, table)
+}
+
+// catchPanic executes the specified function. If a panic occurs, it is returned
+// as an error value.
+func catchPanic(f func() error) (err error) {
+	defer func() {
+		if r := recover(); r != nil {
+			err = r.(error)
+			log.Criticalf("Caught unhandled error: %v", err)
+		}
+	}()
+
+	err = f()
+
+	return
+}
+
+// View opens a database read transaction and executes the function f with the
+// transaction passed as a parameter. After f exits, the transaction is rolled
+// back. If f errors, its error is returned, not a rollback error (if any
+// occur). The passed reset function is called before the start of the
+// transaction and can be used to reset intermediate state. As callers may
+// expect retries of the f closure (depending on the database backend used), the
+// reset function will be called before each retry respectively.
+func (db *db) View(f func(tx walletdb.ReadTx) error, reset func()) error {
+	reset()
+
+	tx, err := newReadWriteTx(db, true)
+	if err != nil {
+		return err
+	}
+
+	err = catchPanic(func() error { return f(tx) })
+	if err != nil {
+		_ = tx.Rollback()
+		return err
+	}
+
+	return tx.Commit()
+}
+
+// Update opens a database read/write transaction and executes the function f
+// with the transaction passed as a parameter. After f exits, if f did not
+// error, the transaction is committed. Otherwise, if f did error, the
+// transaction is rolled back. If the rollback fails, the original error
+// returned by f is still returned. If the commit fails, the commit error is
+// returned. As callers may expect retries of the f closure, the reset function
+// will be called before each retry respectively.
+func (db *db) Update(f func(tx walletdb.ReadWriteTx) error, reset func()) (err error) {
+	reset()
+
+	tx, err := newReadWriteTx(db, false)
+	if err != nil {
+		return err
+	}
+
+	err = catchPanic(func() error { return f(tx) })
+	if err != nil {
+		_ = tx.Rollback()
+		return err
+	}
+
+	return tx.Commit()
+}
+
+// PrintStats returns all collected stats pretty printed into a string.
+func (db *db) PrintStats() string {
+	return ""
```

Comment:
nit: return something like `"<stats not supported by postgres driver>"`?

Reply:
done